### PR TITLE
build: add spec to build RPM

### PR DESCRIPTION
**Pull request by runcom**
Fixes https://github.com/fedora-iot/fido-device-onboard-rs/issues/135

Signed-off-by: Antonio Murdaca <runcom@linux.com>

### CODE REVIEW

File: fido-device-onboard.spec [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+%install
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-client-linuxapp
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-manufacturing-client
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-manufacturing-server
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-owner-onboarding-server
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-rendezvous-server
+# TODO(runcom): we may want to move this to {_bindir} so admins can use it directly
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-owner-tool
+install -D -m 0644 -t %{buildroot}%{_unitdir} examples/systemd/*
+# TODO(runcom): this has to go under /usr (???)
```
Comment (runcom) [inline fido-device-onboard.spec:43]: cross linking my q from https://github.com/fedora-iot/fido-device-onboard-rs/pull/75/files#r732989096

why `/usr/fdo` - rn I'm putting the configs under `/etc/fdo`

File: examples/systemd/fdo-owner-onboarding-server.service [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=FDO owner onboarding server
+After=network-online.target
+
+[Service]
+Environment=LOG_LEVEL=trace
```
Comment (puiterwijk) [inline examples/systemd/fdo-owner-onboarding-server.service:6]: I don't think we should have `LOG_LEVEL=trace` by default. I think this should be `info`. (We have an outstanding issue to check all log levels).

File: fido-device-onboard.spec [hunk 3]
```diff
@@ -0,0 +1,122 @@
+%define dracutlibdir %{_prefix}/lib/dracut
+%bcond_without check
+%global __cargo_skip_build 0
+%global __cargo_is_lib() false
+%global gitversion 43f7674dad4abc470cdab505fbf49f571539cea8
```
Comment (puiterwijk) [inline fido-device-onboard.spec:5]: Is there any chance that the one we add in the repo has the gitversion automatically determined?

File: fido-device-onboard.spec [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+Name:           fido-device-onboard
+Version:        0.1.0
+Release:        1%{?dist}
+Summary:        An implementation of the FIDO Device Onboard Specification written in rust
+
+License:        BSD 3
+URL:            https://github.com/fedora-iot/fido-device-onboard-rs/
+Source:         %{url}/archive/%{gitversion}/%{name}-rs-%{gitversion}.tar.gz
+Source1:        %{name}-rs-%{gitversion}-vendor.tar.gz
```
Comment (puiterwijk) [inline fido-device-onboard.spec:15]: Given that Fedora doesn't vendor things, maybe make this Source dependent on the distro it's being built for?

File: fido-device-onboard.spec [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+%prep
+%autosetup -n %{name}-rs-%{gitversion} -p1
+%cargo_prep -V 1
+
+%build
+%cargo_build
+
+%install
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-client-linuxapp
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-manufacturing-client
```
Comment (puiterwijk) [inline fido-device-onboard.spec:36]: Isn't libexecdir only for binaries that are being executed by other binaries?
These are either executed by hand, or by a systemd unit, and I thought that would have to go into /usr/bin?

File: fido-device-onboard.spec [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-manufacturing-server
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-owner-onboarding-server
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-rendezvous-server
+# TODO(runcom): we may want to move this to {_bindir} so admins can use it directly
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-owner-tool
```
Comment (puiterwijk) [inline fido-device-onboard.spec:41]: This definitely should go into bindir.

File: fido-device-onboard.spec [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+install -D -m 0644 -t %{buildroot}%{_unitdir} examples/systemd/*
+install -D -m 0644 -t %{buildroot}%{_datadir}/fdo examples/config/*
```
Comment (puiterwijk) [inline fido-device-onboard.spec:43]: Shouldn't this be `%{_sysconfdir}` (i.e. `/etc`)?

File: examples/systemd/fdo-owner-onboarding-server.service (hunk 2 above)
Comment (runcom) [reply examples/systemd/fdo-owner-onboarding-server.service:6]: I'll move these to `info`

File: fido-device-onboard.spec (hunk 3 above)
Comment (runcom) [reply fido-device-onboard.spec:5]: not sure I fully understand - this is also just for testing, we're gonna have semver versions here I think

File: fido-device-onboard.spec (hunk 4 above)
Comment (runcom) [reply fido-device-onboard.spec:15]: work to make this spec working on fedora is still ongoing but I don't want to block rhel packaging on that

File: fido-device-onboard.spec (hunk 5 above)
Comment (runcom) [reply fido-device-onboard.spec:36]: not fully sure, that's something I've always done for systemd unit cause those binaries aren't called by users (and systemd isn't a user right?)

File: fido-device-onboard.spec (hunk 6 above)
Comment (runcom) [reply fido-device-onboard.spec:41]: yeep

File: fido-device-onboard.spec (hunk 7 above)
Comment (runcom) [reply fido-device-onboard.spec:43]: we said we wanted these config under `/usr/share` - I'm fine either way, we just have to decide before I change it again :smile:

Comment (runcom) [reply fido-device-onboard.spec:43]: also, those are just examples, not even made for actually starting the services (as you need keys and other stuff anyway)

File: fido-device-onboard.spec (hunk 4 above)
Comment (runcom) [reply fido-device-onboard.spec:17]: it's now rhel only but we still need to find a way to add buildrequires for fedora

File: fido-device-onboard.spec (hunk 3 above)
Comment (runcom) [reply fido-device-onboard.spec:5]: ok I understand - how would you do that tho? :thinking:

Comment (runcom) [reply fido-device-onboard.spec:5]: we can template the spec and add a Makefile target that does that if we're ever going to support a `make rpm` target - ok I'll do that

Comment (runcom) [reply fido-device-onboard.spec]: done, there's a `make rpm` now that builds the RPM and populates the commit variable as well

File: fido-device-onboard.spec (hunk 7 above)
Comment (puiterwijk) [reply fido-device-onboard.spec:59]: Ah, right. In that case, maybe they should go under `%{_docdir}`?

Comment (nullr0ute) [reply fido-device-onboard.spec:59]: Yes

File: dracut/52fdo/manufacturing-client.service [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+Before=coreos-installer-reboot.service
+ConditionPathExists=/etc/manufacturing-client-config
+Requires=dev-disk-by\x2dlabel-boot.device
+
+OnFailure=emergency.target
+OnFailureJobMode=replace-irreversibly
+
+[Service]
+Type=oneshot
+EnvironmentFile=/etc/manufacturing-client-config
```
Comment (puiterwijk) [inline dracut/52fdo/manufacturing-client.service:15]: Should we add `LOG_LEVEL=info` here as well?
Does systemd just append that to the contents it reads from the EnvironmentFile?

File: fido-device-onboard.spec (hunk 7 above)
Comment (runcom) [reply fido-device-onboard.spec:59]: so `%{_docdir} it is

File: examples/config/rendezvous-info.yml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+- dns: fdo-test.redhat.com
```
Comment (puiterwijk) [inline examples/config/rendezvous-info.yml:2]: Let's make this generic? Something like `fdo.example.com`?

File: examples/config/owner-addresses.yml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+- transport: HTTP
+  port: 8081
+  addresses:
+    - dns_name: fdo-test.redhat.com
```
Comment (puiterwijk) [inline examples/config/owner-addresses.yml:5]: Let's make this e.g. `fdo.example.com`?

**Comment by runcom** [comment]
Should really be ready now? @puiterwijk @nullr0ute @jnogol ptal - I want to get this built asap (at least in RHEL) and we can update whatever we want as follow up :)

**Comment by nullr0ute** [comment]
Looking

**Comment by runcom** [comment]
The TL;DR; for Fedora is that we cannot use automatic Requires/BuildRequires as this project uses a workspace and macros aren't supporting that - again, I think this is a great start for packaging and unblocks us from building in RHEL and by using `make rpm` you can also jump on a RHEL machine and build local packages anyway

**Comment by nullr0ute** [comment]
Can we not just smash all of of this into a single commit?

**Comment by runcom** [comment]
> Can we not just smash all of of this into a single commit?

not sure what you mean :thinking: I've split changes into commit as you requested above - something might have slipped but can you clarify?

File: fido-device-onboard.spec [hunk 11]
```diff
@@ -0,0 +1,133 @@
+%define dracutlibdir %{_prefix}/lib/dracut
+%bcond_without check
+%global __cargo_skip_build 0
+%global __cargo_is_lib() false
+%global forgeurl https://github.com/fedora-iot/fido-device-onboard-rs
```
Comment (nullr0ute) [inline fido-device-onboard.spec:5]: It doesn't look like this or the forgemeta below are used anywhere

File: fido-device-onboard.spec [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+%forgemeta
+
+Name:           fido-device-onboard
+Version:        0.1.0
+Release:        1%{?dist}
+Summary:        An implementation of the FIDO Device Onboard Specification written in rust
+
+License:        BSD 3
+URL:            https://github.com/fedora-iot/fido-device-onboard-rs
+Source:         %{url}/archive/%{commit}/%{name}-rs-%{commit}.tar.gz
```
Comment (nullr0ute) [inline fido-device-onboard.spec:16]: Commit isn't defined anywhere, we should probably just use the version nomenclature everywhere

File: fido-device-onboard.spec [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-client-linuxapp
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-manufacturing-client
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-manufacturing-server
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-owner-onboarding-server
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-rendezvous-server
+install -D -m 0755 -t %{buildroot}%{_bindir}/fdo target/release/fdo-owner-tool
+install -D -m 0644 -t %{buildroot}%{_unitdir} examples/systemd/*
+install -D -m 0644 -t %{buildroot}%{_docdir}/fdo examples/config/*
+# 52fdo
+install -D -m 0755 -t %{buildroot}%{dracutlibdir}/modules.d/52fdo dracut/52fdo/module-setup.sh
```
Comment (nullr0ute) [inline fido-device-onboard.spec:56]: I think this should just be called 52fdo IE drop the dracut in the module name as it's implied by the location.

File: fido-device-onboard.spec [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+%files -n fdo-owner-onboarding-server
+%license LICENSE
+%{_libexecdir}/fdo/fdo-owner-onboarding-server
+%{_docdir}/fdo/owner-onboarding-server.yml
+%{_unitdir}/fdo-owner-onboarding-server.service
+
+%package -n fdo-rendezvous-server
+Summary: FDO Rendezvous Server implementation
+License: BSD 3
```
Comment (nullr0ute) [inline fido-device-onboard.spec:86]: Can drop all the License lines from the subpackages, if not there they inherit from the root package. Note the %license lines need to stay

**Review by nullr0ute — CHANGES_REQUESTED** [review]

File: fido-device-onboard.spec (hunk 11 above)
Comment (runcom) [reply fido-device-onboard.spec:5]: they're used to generate metadata as per fedora/rhel guidelines

**Comment by nullr0ute** [comment]
> > Can we not just smash all of of this into a single commit?
> 
> not sure what you mean thinking I've split changes into commit as you requested above - something might have slipped but can you clarify?

Ignore me, I need coffee...

File: fido-device-onboard.spec (hunk 12 above)
Comment (runcom) [reply fido-device-onboard.spec:16]: the Makefile defines it - downstream will add it, this is the very same osbuild/osbuild-composer does so you can just git clone this and build a fully functional RPM

File: fido-device-onboard.spec (hunk 13 above)
Comment (runcom) [reply fido-device-onboard.spec:56]: will do perfect :+1:

**Comment by runcom** [comment]
> Ignore me, I need coffee...

we ALL need coffee :coffee:

File: fido-device-onboard.spec (hunk 14 above)
Comment (runcom) [reply fido-device-onboard.spec:86]: awesome

File: fido-device-onboard.spec (hunk 13 above)
Comment (runcom) [reply fido-device-onboard.spec:56]: there's no `dracut` in the name :thinking:

File: fido-device-onboard.spec (hunk 14 above)
Comment (runcom) [reply fido-device-onboard.spec:86]: done

**Comment by runcom** [comment]
@nullr0ute all comments addressed

File: Makefile [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+RPM_SPECFILE=rpmbuild/SPECS/fido-device-onboard-rs-$(COMMIT).spec
+RPM_TARBALL=rpmbuild/SOURCES/fido-device-onboard-rs-$(COMMIT).tar.gz
+VENDOR_TARBALL=rpmbuild/SOURCES/fido-device-onboard-rs-$(COMMIT)-vendor.tar.gz
+
+$(RPM_SPECFILE):
+	mkdir -p $(CURDIR)/rpmbuild/SPECS
+	(echo "%global gitversion $(COMMIT)"; git show HEAD:fido-device-onboard.spec) > $(RPM_SPECFILE)
+
+$(RPM_TARBALL):
+	mkdir -p $(CURDIR)/rpmbuild/SOURCES
```
Comment (nullr0ute) [inline Makefile:47]: Should we add the cargo vendor and bits to make the tar file here too?

Comment (runcom) [reply Makefile:47]: cargo vendor it's right below this

Comment (runcom) [reply Makefile:47]: reason for the split is that fedora doesn't need vendor so when we support fedora in full, we add a condition on fedora itself

File: fido-device-onboard.spec (hunk 13 above)
Comment (nullr0ute) [reply fido-device-onboard.spec:56]: It doesn't read like that here (or in the downloaded spec), and it looks like it has a space. copy from my local spec is ````%{dracutlibdir}/modules.d/52fdo dracut/52fdo/````:

dracut

File: fido-device-onboard.spec [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+%endif
+BuildRequires: systemd
+
+%description
+%{summary}.
+
+%prep
+%autosetup -n %{name}-rs-%{commit} -p1
+%if 0%{?rhel} && !0%{?eln}
+%cargo_prep -V 1
```
Comment (nullr0ute) [inline fido-device-onboard.spec:38]: Is there a reason we need to -V 1 here and not elsewhere or the reverse?

File: fido-device-onboard.spec [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+install -D -m 0755 -t %{buildroot}%{_libexecdir}/fdo target/release/fdo-rendezvous-server
+install -D -m 0755 -t %{buildroot}%{_bindir}/fdo target/release/fdo-owner-tool
+install -D -m 0644 -t %{buildroot}%{_unitdir} examples/systemd/*
+install -D -m 0644 -t %{buildroot}%{_docdir}/fdo examples/config/*
+# 52fdo
+install -D -m 0755 -t %{buildroot}%{dracutlibdir}/modules.d/52fdo dracut/52fdo/module-setup.sh
+install -D -m 0755 -t %{buildroot}%{dracutlibdir}/modules.d/52fdo dracut/52fdo/manufacturing-client-generator
+install -D -m 0755 -t %{buildroot}%{dracutlibdir}/modules.d/52fdo dracut/52fdo/manufacturing-client-service
+install -D -m 0755 -t %{buildroot}%{dracutlibdir}/modules.d/52fdo dracut/52fdo/manufacturing-client.service
+
```
Comment (nullr0ute) [inline fido-device-onboard.spec:60]: systemd services need the following lines for each sub package that ships services. add a -n XXX like in the %files for each %XXX

%post
%systemd_post blah.service

%preun
%systemd_preun blah.service

%postun
%systemd_postun_with_restart blah.service

**Comment by nullr0ute** [comment]
Example systemd bits with a rust service here:
https://src.fedoraproject.org/rpms/parsec/blob/rawhide/f/parsec.spec

File: fido-device-onboard.spec (hunk 16 above)
Comment (runcom) [reply fido-device-onboard.spec:38]: this is just for RHEL to use vendored deps

File: fido-device-onboard.spec [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+%files -n fdo-owner-onboarding-server
+%license LICENSE
+%{_libexecdir}/fdo/fdo-owner-onboarding-server
+%{_docdir}/fdo/owner-onboarding-server.yml
+%{_unitdir}/fdo-owner-onboarding-server.service
+
+%package -n fdo-rendezvous-server
+Summary: FDO Rendezvous Server implementation
+%description -n fdo-rendezvous-server
```
[the remark this answers is not in the document]
Comment (runcom) [reply fido-device-onboard.spec:84]:
```
If package scriptlets call other systemd tools, for example systemd-tmpfiles, the package SHOULD declare appropriate dependencies. The %{?systemd_requires} macro is a shortcut to require systemd for the %pre, %post, and %postun scriptlets. Note that those dependencies are not required for the %systemd_{post,preun,postun_with_restart,user_post,user_preun} macros listed above.
```
I think we don't need this but just the macro you suggested below

File: fido-device-onboard.spec (hunk 13 above)
Comment (runcom) [reply fido-device-onboard.spec:56]: I don't know where and how you're getting that :joy: the module is called 52fdo and in my testing it's indeed fdo to pull it in in dracut

Comment (runcom) [reply fido-device-onboard.spec:56]: aahhhh that's just the `install` syntax - the space is for install TARGET SRC so doesn't have anything to do with the name too

File: fido-device-onboard.spec (hunk 18 above)
[the remark this answers is not in the document]
Comment (nullr0ute) [reply fido-device-onboard.spec:84]: ````The %{?systemd_requires} macro is a shortcut to require systemd for the %pre, %post, and %postun scriptlets.````

The sentence above states it's a shortcut to require systemd for rpm scriptlets, then you say it's not needed in the following sentence below.

````Note that those dependencies are not required for the %systemd_{post,preun,postun_with_restart,user_post,user_preun} macros listed above.````

File: fido-device-onboard.spec (hunk 17 above)
Comment (runcom) [reply fido-device-onboard.spec:60]: done, notice the manufacturing-client/dracut module don't need this as this is just one-shot in the initramfs

**Review by nullr0ute — APPROVED** [review]
LGTM. We can always tweak it as necessary later

File: fido-device-onboard.spec (hunk 18 above)
[the remark this answers is not in the document]
Comment (runcom) [reply fido-device-onboard.spec:93]: > then you say it's not needed in the following sentence below.

those are fedora guidelines not me :smile: and given the above we don't need it as we don't require systemd in pre/post/postun

**Pull request merged**